src/Lean/Meta/ExprDefEq.lean

```diff
@@ -1963,11 +1963,14 @@ where
     let sFn := s.getAppFn
     if !sFn.isMVar then
       return false
-    if (← isAssignable sFn) then
-      let ctorApp := mkApp (mkAppN (mkConst ctorVal.name sTypeFn.constLevels!) sType.getAppArgs) v
-      processAssignment' s ctorApp
-    else
+    if !(← isAssignable sFn) then
+      return false
+    let ctor := mkAppN (mkConst ctorVal.name sTypeFn.constLevels!) sType.getAppArgs
```

Comment:
Could you please add a comment explaining this change? It can be based on the PR description. 
 Additional requests regarding coding convetion:
 - `let Expr.forallE _ ty ...` => `let .forallE _ ty ...`
 - `unless ← isDefEq ty (← inferType v) do` => `unless (← isDefEq ty (← inferType v)) do`

Reply:
Done.

General question about `isDefEq` that came up in a Zulip discussion: some users believe that you are not supposed to use `isDefEq` on terms if you have not already ensured `isDefEq` for their inferred types. Is that the case?

If so, then this PR is not necessary, but it would raise a bunch of questions for me (I don't recall seeing any meta code that specifically ensured `isDefEq` on inferred types).

Reply:
I don't have time to context switch to this module right now. Moreover, I feel like this is going to be a really long thread. Let's create one at the Lean FRO Zulip. 

BTW, I want to minimize as much as possible changes to `ExprDefEq.lean`, and do it in  batches whenever we decide to the context switch.

That said, at first sight, this PR is consistent with the current design. We already have functions such as `checkTypesAndAssign`.

Reply:
BTW, were you able to trigger the bug without meta-programming? If not, I prefer to put this PR on hold.

Reply:
No problem, it's low-priority. I'll let you know if I manage to trigger it without meta-programming.

Reply:
Just to let you know, I pushed a simple example that uses just the `have` and `exact` tactics. (Please feel free to keep this PR on hold.)